File: directMethods.py
import numpy as np
import itertools as itt
"""
Classe DirectMethods

Solução de sistemas lineares via métodos diretos
"""
class DirectMethods:

  def progressive_substituition(L, b):
    """ Obter o vetor x tal que torne verdadeira a equação Lx = b

    L : ndarray
      matriz triangular inferior 2D com tipos float
    b : ndarray
      vetor de resultado 1D com tipos float
    """
    n = len(L)

    x = []
    for i in range(n):
      x_i = b[i]

      for j in range(i):
        x_i -= L[i][j] * x[j]

      x_i /= L[i][i]
      x.append(x_i)
    
    return np.array(x)

  def regressive_substitution(U, b):
    """ Obter o vetor x tal que torne verdadeira a equação Ux = b

    U : ndarray
      matriz triangular superior 2D com tipos float
    b : ndarray
      vetor de resultado 1D com tipos float
    """
    n = len(U)

    x = [0]*n
    for i in range(n-1,-1,-1):
      x_i = b[i]

      for j in range(i, n):
        x_i -= U[i][j] * x[j]

      x_i /= U[i][i]
      x[i] = x_i

    return np.array(x)

  def decomposition_lu(A):
    """ Obter o vetor decomposição LU da matriz A tal que A = LU

    A : ndarray
      2D array com tipos float
    """
    n = len(A)

    L = np.identity(n)
    U = np.zeros((n,n))

    for i in range(n):
      for j in range(n):

        if i <= j:
          sumTerm = np.sum(L[i,:i] * U[:i,j])
          U[i,j] = A[i][j] - sumTerm

        if i > j:
          sumTerm = np.sum(L[i,:j] * U[:j,j])
          L[i,j] = (A[i,j] - sumTerm) / U[j,j]

    return {'L': L, 'U': U}

  def decomposition_cholesky(A):
    """ Obter o vetor decomposição H da matriz A tal que A = H(H^T) via Cholesky

    A : ndarray
      2D array com tipos float
    """
    n = len(A)
    H = np.zeros((n,n))

    for j in range(n):
      h_ij = 0
      for i in range(j,n):

        if i == j:
          h_ij = np.sqrt(A[i,j] - sum(H[i,:i]**2))
        else:
          h_ij = (A[i,j] - sum(H[i,:i]*H[j,:i])) / H[j,j]

        H[i,j] = h_ij

    return H

  def gauss_elimination(A, b):
    """ Obter a matriz triangular superior de A via eliminação de Gauss

    A : ndarray
      2D array com tipos float
    b : ndarray
      vetor de resultado 1D com tipos float
    """
    n = len(A)

    for j in range(n):
      # Com pivot
      pivot = max(abs(A[j:,j]))
      # indice da linha que contem 'pivot' em seu triangulo inf
      p = [k for k in range(j, n) if pivot == A[k,j]]
      if not p: continue
      p = p[0]
      temp = np.array(A[p,:])
      A[p,:] = A[j,:]
      A[j,:] = temp

      temp = b[j]
      b[j] = b[p]
      b[p] = temp

      # print(A[j,j])
      for i in range(j+1,n):
        m_ij = - A[i,j] / A[j,j]
        A[i,j:] = A[i,j:] + m_ij * A[j,j:]

        b[i] = b[i] + m_ij * b[j]

    return {'A': A, 'b': b}
  
  def inverse(A):
    """ Obter a matriz inversa de A (A^(-1))

    A : ndarray
      2D array com tipos float
    """
    n = len(A)
    I = np.eye(n)

    v = []

    LU = DirectMethods.decomposition_lu(A)

    for e in I:
      y = DirectMethods.progressive_substituition(LU['L'], e)
      x = DirectMethods.regressive_substitution(LU['U'], y)

      v.append(x)

    return np.array(v).transpose()

File: directMethods.js
function transpose(A) {
  return A.map((row,i) => row.map((_,j) => A[j][i]));
}

function matmul(A,B) {
  const BT = transpose(B)
  return A.map((row, i) => row.map((_, j) => row.reduce((s, _, k) => s + A[i][k]*BT[j][k], 0)))
}

function progressive_substitution(L, b) {
  const N = L.length

  const Z = Array(N).fill(0)
  const X = Z.map((_, i, V) => {
    
    // let formula = `x_${i} = `
    // formula += `[b_${i} - `
    // L[i].reduce((s, _, j) => i > j ? formula += `L_${i}${j}*x_${j} + ` : '', 0)
    // formula += `] / a_${i}${i}`
    // console.log(formula)

    const sumTerm = L[i].reduce((s, _, j) => i > j ? s + L[i][j]*V[j] : s, 0)
    V[i] = (b[i] - sumTerm) / L[i][i]

    return V[i]
  })

  return X;
}

function regressive_substitution(U, b) {
  const N = U.length

  const UT = U.map((row, i, U) => row.map((_,j) => i >= j ? U[j][i] : 0))
  const bRev = b.map((_, i, V) => V[V.length-1-i]);

  const xRev = progressive_substitution(UT, bRev)
  const X = xRev.map((_, i, V) => V[b.length-1-i]);
  return X
}

function decomposition_cholesky(A) {
  const N = A.length
  const Z = [...Array(N)].map(_ => Array(N).fill(0));

  const H = Z.map((row, i, mat) => row.map((a, j) => {
    let sumTerm;

    // let formula = `h_${i}${j} = `
    if (i == j) {
      // formula += `sqrt[`
      // formula += `a_${i}${j} - `
      // row.reduce((s, _, k) => k < j ? formula += `(h_${i}${k})^2 +` : s, 0)
      // formula += `]`
      // console.log(formula)

      sumTerm = row.reduce((s, _, k) => k < j ? s + mat[i][k] * mat[j][k] : s, 0)
      mat[i][j] = Math.sqrt(A[i][j] - sumTerm)
      return mat[i][j]
    }

    if (j < i ) {
      // formula += `[`
      // formula += `a_${i}${j} - `
      // row.reduce((s, _, k) => k < j ? formula += `(h_${i}${k})^2 +` : s, 0)
      // formula += `] / h_${j}${j}`
      // console.log(formula)

      sumTerm = row.reduce((s, _, k) => k < j ? s + mat[i][k] * mat[j][k] : s, 0)
      mat[i][j] = (A[i][j] - sumTerm) / mat[j][j]
      return mat[i][j]
    }
    
    // formula += '0'
    // console.log(formula)

    mat[i][j] = a
    return mat[i][j]
  }))


  return H
}

function decomposition_LU(A) {
  const N = A.length

  const L = [...Array(N)].map(() => Array(N).fill(0))
  const U = [...Array(N)].map(() => Array(N).fill(0))

  L.map((_, i) => L[i][i] = 1)
  A.map((row, i, M) => row.map((v, j) => {

    if (i <= j) {
      const sumTerm = row.reduce((s, _, k) => k < i ? s + L[i][k]*U[k][j] : s, 0)
      const u_ij = A[i][j] - sumTerm
      U[i][j] = u_ij
      return U[i][j]
    }

    if (i > j) {
      const sumTerm = row.reduce((s, _, k) => k < j ? L[i][k]*U[k][j] : s, 0)
      const l_ij = (A[i][j] - sumTerm) / U[j][j]
      L[i][j] = l_ij
      return L[i][j]
    }
  }))

  return {L, U}
}

function gauss_elimination_matrix(A) {

  const N = A.length

  A.map((row, i) => row.map((_, j) => {
    const k = j + i + 1
    if (k >= N) return

    const m_ij = - A[k][i] / A[i][i]
    row.map((_, l) => A[k][l] = A[k][l] + (m_ij * A[i][l]))

  }))
  return A
}



// const L = mathjs.matrix([[1,0,0],[10,20,0],[100,200,300]])
// const U = mathjs.matrix([[100,200,300], [0,20,10], [0,0,20]])

// const x = progressive_substitution(L, b)
// const x = regressive_substitution(U, b)
const A = [
  [2, 1, 2],
  [4, 3, 3],
  [6, 5, -1],
]

const L = [
  [10, 0, 0],
  [5, 3, 0],
  [2, 2, 3],
]

const U = [
  [10, 5, 2],
  [0, 3, 2],
  [0, 0, 3],
]
const b = [10, 20, 4]

// console.log(progressive_substitution(L, b))
// console.log(regressive_substitution(U, b))
// console.log(decomposition_cholesky(A))
// console.log(decomposition_LU(A))
console.log(gauss_elimination_matrix(A, b))

